src/transformers/models/mistral/modeling_tf_mistral.py

```diff
@@ -0,0 +1,985 @@
+
+        if (self.head_dim * self.num_heads) != self.hidden_size:
+            raise ValueError(
+                f"hidden_size must be divisible by num_heads (got `hidden_size`: {self.hidden_size}"
+                f" and `num_heads`: {self.num_heads})."
+            )
+        self.q_proj = tf.keras.layers.Dense(self.num_heads * self.head_dim, use_bias=False, name="q_proj")
+        self.k_proj = tf.keras.layers.Dense(self.num_key_value_heads * self.head_dim, use_bias=False, name="k_proj")
+        self.v_proj = tf.keras.layers.Dense(self.num_key_value_heads * self.head_dim, use_bias=False, name="v_proj")
+        self.o_proj = tf.keras.layers.Dense(self.hidden_size, use_bias=False, name="o_proj")
+
+        self.rotary_emb = TFMistralRotaryEmbedding(
+            self.head_dim,
+            max_position_embeddings=self.max_position_embeddings,
+            base=self.rope_theta,
+            name="rotary_emb",
+        )
+
+    def _shape(self, tensor: tf.Tensor, seq_len: int, bsz: int):
+        tensor = tf.reshape(tensor, (bsz, seq_len, self.num_heads, self.head_dim))
+        tensor = tf.transpose(tensor, perm=(0, 2, 1, 3))
+        return tensor
+
+    def call(
+        self,
+        hidden_states: tf.Tensor,
+        attention_mask: Optional[tf.Tensor] = None,
+        position_ids: Optional[tf.Tensor] = None,
+        past_key_value=None,
+        output_attentions: bool = False,
+        use_cache: bool = False,
+        **kwargs,
+    ) -> Tuple[tf.Tensor, Optional[tf.Tensor], Optional[Tuple[tf.Tensor]]]:
+        if "padding_mask" in kwargs:
+            warnings.warn(
+                "Passing `padding_mask` is deprecated and will be removed in v4.37. Please make sure use `attention_mask` instead.`"
+            )
+        bsz, q_len, _ = shape_list(hidden_states)
+
+        query_states = self.q_proj(hidden_states)
+        key_states = self.k_proj(hidden_states)
+        value_states = self.v_proj(hidden_states)
+
+        query_states = tf.transpose(
+            tf.reshape(query_states, (bsz, q_len, self.num_heads, self.head_dim)), perm=(0, 2, 1, 3)
+        )
+        key_states = tf.transpose(
+            tf.reshape(key_states, (bsz, q_len, self.num_key_value_heads, self.head_dim)), perm=(0, 2, 1, 3)
+        )
+        value_states = tf.transpose(
+            tf.reshape(value_states, (bsz, q_len, self.num_key_value_heads, self.head_dim)), perm=(0, 2, 1, 3)
+        )
+
+        kv_seq_len = shape_list(key_states)[-2]
+        if past_key_value is not None:
+            if self.layer_idx is None:
+                raise ValueError(
+                    f"The cache structure has changed since version v4.36. If you are using {self.__class__.__name__} "
+                    "for auto-regressive decoding with k/v caching, please make sure to initialize the attention class "
+                    "with a layer index."
+                )
+            kv_seq_len += past_key_value.get_usable_length(kv_seq_len, self.layer_idx)
```

Comment:
@Rocketknight1 how do we deal with Cache in TensorFlow?

I was unable to find a one to one correspondance to any models -- apologies if my search was not thorough. If you could refer me to an implementation I could read that up and make changes accordingly.

Reply:
Hey! Unfortunately, this bit is unfortunately going to be trickier. The reason is that our newer Torch models have transitioned to a new `Cache` structure, but our TF models (as well as a lot of older Torch models) still use a structure like `Tuple[Tuple[tf.Tensor]]` or `Tuple[Tuple[torch.Tensor]]`. 

Because our TF generation code is still using the old structure (for now - we need to fix this!), you'll have to avoid `Cache` and `DynamicCache` and instead base the past key-values on the implementation from another TF model. It's definitely one of the most annoying bits, so if you need me to handle this part of the port, let me know!

Reply:
Understandable!

So there are two options as I see:
1. If you could point me to another TF model that solves this, I might be able to understand this bit better.
2. If you think there would be a lot of "to and fro" for me to implement this, and it would be effective for you to implement, please feel free to handle this part of the port.